sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -908,21 +908,35 @@ object ScalaJSPluginInternal {
       scalaJSModuleInitializers :=
         (scalaJSModuleInitializers in (This, Zero, This)).value,
 
-      scalaJSModuleInitializers ++= {
+      scalaJSModuleInitializers ++= Def.taskDyn {
```

Comment:
Is that `taskDyn` really necessary? Forcing evaluation of `scalaJSDiscoveredMainClasses` even when `scalaJSUseMainModuleInitializer` is false shouldn't be an issue, I think.

If it's not necessary I would recommend not to use `taskDyn`, as each one deteriorates the debuggability of builds through things like `inspect tree`.

It is is absolutely necessary for correctness, at least use `Def.settingDyn` instead. This is possible because the dynamic part relies only on `scalaJSUseMainModuleInitializer` which is a setting.

Reply:
Honestly, I was unfamiliar difference between `taskDyn` and `settingDyn`.
Fixed in https://github.com/scala-js/scala-js/pull/3724/commits/3cec2d48f6bdff829ea1202a3f87f0f719b28304

Reply:
Did you try without using any `Dyn` at all?

Reply:
Yes, tried in https://github.com/scala-js/scala-js/commit/d6e9944923dff56411d179d3fd38cb86bd7661a3
But it turned out that `Dyn` is required since the referenced `scalaJSDiscoveredMainClasses` are `TaskKey`.

Reply:
Hum, what do you mean? Was it just that the macro complained that you can't access a task key's value in an if or lambda? Because that can be dealt with easily by storing `scalaJSDiscoveredMainClasses.value` in a `val` first, before doing the test `if (scalaJSUseMainModuleInitializer.value) {`. That's the normal way to resolve this issue.

Using Dyns is an entirely different kind of monster, that must be used only when the tasks accessed in the dynamic parts *must not* be evaluated when the test fails. It should be very rare. Unfortunately in 0.6.x the plugin has many of them because we took bad design decisions, but if you look at the 1.x plugin, I believe there is only one: the thing that decides whether to evaluate `fastOptJS` or `fullOptJS` based on `scalaJSStage`. That one is of course critical, because if not done properly, every invocation of `test` would evaluate `fullOptJS`! Here we don't care whether `scalaJSDiscoveredMainClasses` is evaluated even if it's not used.

Reply:
> Was it just that the macro complained that you can't access a task key's value in an if or lambda?

Yes, I think it was.
Removed `Dyn` in https://github.com/scala-js/scala-js/pull/3724/commits/11e8dc52c613ee8cab13bb7f2df8f2053dc4b7d3